server/build_event_protocol/build_event_handler/build_event_handler.go

```diff
@@ -295,8 +306,17 @@ func (e *EventChannel) handleEvent(event *pepb.PublishBuildToolEventStreamReques
 		if err := e.env.GetInvocationDB().InsertOrUpdateInvocation(e.ctx, ti); err != nil {
 			return err
 		}
+	} else if !e.hasReceivedStartedEvent {
+		e.eventsBeforeStarted = append(e.eventsBeforeStarted, event)
+		if len(e.eventsBeforeStarted) > 10 {
+			log.Printf("We got over 10 build events before the started event for invocation %s, dropping %+v", iid, e.eventsBeforeStarted[0])
+			e.eventsBeforeStarted = e.eventsBeforeStarted[1:]
+		}
+		return nil
 	}
 
+	e.beValues.AddEvent(&bazelBuildEvent) // in-memory structure to hold common values we want from the event.
```

Comment:
all of the code from 318:352 is called (eventually) on every event -- could we pull that out to a new method on this function, and then this outer function can do away with the recursive stuff and be:

func processSingleEvent(event) {
 <code from 318:352>
}

handleEvent() {
  if isStartedEvent() {
    do_started_stuff()
  } else if !receivedStartedEvent() {
    buffer_event()
    return
  }

  // Process buffered events.
  for _, event := range eventsBeforeStarted{
    processSingleEvent(event)
  }
  eventsBeforeStarted = nil

  // Process regular events.
 processSingleEvent(event)
}

Reply:
Done